docs/core/upgrade.md

```diff
@@ -0,0 +1,108 @@
+
+## Tracking Module Versions
+
+Each module gets assigned a consensus version by the module developer, which serves as the breaking change version of the module. The SDK keeps track of all modules' consensus versions in the x/upgrade's `VersionMap` store. During an upgrade, the SDK calculates the difference between the old `VersionMap` stored in state and the new `VersionMap`. For each difference found, the SDK will run module-specific migrations and increment the respective consensus version of each upgraded module.
+
+The next paragraphs detail each component of the in-place store migration process, and gives instructions on how to update your app to take advantage of this functionality.
+
+## Genesis State
+
+When starting a new chain, each module's consensus version must be saved to state during the application's genesis. This can be done by adding the following line to the `InitChainer` method in `app.go`
+
+```diff
+func (app *MyApp) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.ResponseInitChain {
+  ...
++ app.UpgradeKeeper.SetModuleVersionMap(ctx, app.mm.GetVersionMap())
+  ...
+}
+```
+
+Using this information, the SDK will be able to detect when modules with newer versions are introduced to the app. 
+
+### Consensus Version
+The consensus version is defined on each app module by the module developer. It serves as the breaking change version of the module. The consensus version helps to inform the SDK on which modules need to be upgraded. For example, if the bank module was version 2 and an upgrade introduces bank module 3, the SDK will know to upgrade the bank module and run its "version 2 to 3" migration script.
+
+### Version Map
+The version map is a mapping of module names to consensus versions. The map is persisted to x/upgrade's state for use during in-place migrations. When migrations finish, the updated version map is persisted to state. 
+
+## Upgrade Handlers
+
+Upgrades utilize an `UpgradeHandler` to facilitate migrations. `UpgradeHandler`s are functions implemented by the app developer that conform to the following function signature. These functions retrieve the `VersionMap` from x/upgrade's state, and return the new `VersionMap` to be stored in x/upgrade after the upgrade. The diff between the two `VersionMap`s determines which modules need upgrading.
+
+```golang
+type UpgradeHandler func(ctx sdk.Context, plan Plan, fromVM VersionMap) (VersionMap, error)
+```
+
+Inside these functions, you should perform any upgrade logic you wish to include in the provided `plan`. All upgrade handler functions should end with the following line of code:
```

Comment:
https://developers.google.com/style/word-list#letter-s 
I think we mean "must" here 
Using "should" is almost never the right word. 
> Because should is ambiguous by definition, it can be problematic. For instance, if you're telling the reader what to do, should implies that the action is recommended but optional, leaving the reader unsure about what to do.
Clarify what you mean. Determine if an action is required versus optional, an outcome is expected versus possible, or a state is actual versus recommended.
An action is required. Use must, or rephrase the sentence so that it is a clear imperative instruction such as Do the following before you continue.
An action is recommended. Use We recommend... or Google recommends…. You can use should if a recommended action is generally recognized—for example, You should use a strong password…, or, You should follow the principle of least privilege....